sdk/src/types/block/mana/structure.rs

```diff
@@ -65,3 +108,106 @@ impl Default for ManaStructure {
+    /// `slot_index_target` and applies the decay to the result
+    pub fn potential_mana(
+        &self,
+        amount: u64,
+        slot_index_created: impl Into<SlotIndex>,
+        slot_index_target: impl Into<SlotIndex>,
+    ) -> u64 {
+        let (slot_index_created, slot_index_target) = (slot_index_created.into(), slot_index_target.into());
+        if slot_index_created >= slot_index_target {
+            return 0;
+        }
+        let slots_per_epoch_exp = self.slots_per_epoch_exponent();
+        let mana_structure = self.mana_structure();
+        let (epoch_index_created, epoch_index_target) = (
+            slot_index_created.to_epoch_index(slots_per_epoch_exp),
+            slot_index_target.to_epoch_index(slots_per_epoch_exp),
+        );
+        if epoch_index_created == epoch_index_target {
+            mana_structure.generate_mana(amount, slot_index_target.0 - slot_index_created.0)
+        } else if epoch_index_created == epoch_index_target - 1 {
+            let slots_before_next_epoch =
+                slot_index_created - (epoch_index_created + 1).first_slot_index(slots_per_epoch_exp);
+            let slots_since_epoch_start =
+                slot_index_target - (epoch_index_target - 1).last_slot_index(slots_per_epoch_exp);
+            let mana_decayed = mana_structure.decay(mana_structure.generate_mana(amount, slots_before_next_epoch.0), 1);
+            let mana_generated = mana_structure.generate_mana(amount, slots_since_epoch_start.0);
+            mana_decayed + mana_generated
+        } else {
+            let c = fixed_point_multiply(
+                amount,
+                mana_structure.decay_factor_epochs_sum() * mana_structure.generation_rate() as u32,
+                mana_structure.decay_factor_epochs_sum_exponent() + mana_structure.generation_rate_exponent()
+                    - slots_per_epoch_exp,
+            );
+            let slots_before_next_epoch =
+                slot_index_created - (epoch_index_created + 1).first_slot_index(slots_per_epoch_exp);
+            let slots_since_epoch_start =
+                slot_index_target - (epoch_index_target - 1).last_slot_index(slots_per_epoch_exp);
+            let potential_mana_n = mana_structure.decay(
+                mana_structure.generate_mana(amount, slots_before_next_epoch.0),
+                epoch_index_target.0 - epoch_index_created.0,
+            );
+            let potential_mana_n_1 = mana_structure.decay(c, epoch_index_target.0 - epoch_index_created.0);
+            let potential_mana_0 = c + mana_structure.generate_mana(amount, slots_since_epoch_start.0)
+                - (c >> mana_structure.generation_rate_exponent());
+            potential_mana_0 - potential_mana_n_1 + potential_mana_n
+        }
+    }
+}
+
+/// Returns the upper 32 bits of a u64 value.
+fn upper_bits(v: u64) -> u64 {
+    v >> 32
+}
+
+/// Returns the lower n bits of a u64 value.
+fn lower_n_bits(v: u64, n: u8) -> u64 {
+    debug_assert!(n <= 64);
```

Comment:
```suggestion
    debug_assert!(n > 0 && n <= 64);
```

Reply:
Yea this is needed, ran into it on my local test with prot_params::default(), we gotta get better values for manastruct

Reply:
I don't think this is what we want. it's possible to get the lowest 0 bits which is always 0